### PR TITLE
Refactor pybigtools `values()` rasterization

### LINKED ISSUE #94: Adding more summary options to values binning
Hi, I've been a happy user of pybigtools and specifically `BBIRead.values()` for a while. Elsewhere in our package we use `average_over_bed`, and this has a number of useful summary statistics that aren't available in `values()`. 
The primary one I'd be interested in is simply `sum`. I've been assuming `values`' mean works like mean0 by default if you set `missing` to 0., but if not having a separate `mean0` next to `mean` would also be great. Is it possible to easily add these binning options `values` as well?

### PR DESCRIPTION
**Pull request by nvictus**
This PR reimplements the binned/base-level rasterization in `pybigtools` around a cleaner API, with explicit semantics distinguishing the **inclusion of uncovered bases** (resolves #94) in summary statistic calculations and the post-rasterization **filling of empty bins or positions**. Internally, it also introduces a clean sweepline-based iterator for generating depth-of-coverage intervals from BigBed entries (the prerequisite for BigBed rasterization) and exposes the full set of internal summary stats through `BBIReader.values()`.

## 1. Uncovered-base inclusion/exclusion in statistics and postfill
Previously, the `missing` argument served the purpose of post-rasterization fill-in but not inclusion of uncovered bases in summary calculations, and it's naming was ambiguous.

There is demand for summary statistics that consider uncovered bases as zeros (summarize over the entire bin) rather than only over covered bases, which is not yet exposed via `values()` #94. UCSC exposes some of these with names like `mean0` in averageOverBed but this is all somewhat obscure and many people don't even realize that plain `mean` excludes uncovered bases.

The goal here is to make both functionality possible via two arguments:
* `uncovered: float | None` -- determines whether uncovered bases are included in statistics. By default (`None`) they are excluded and any empty positions or summary bins are returned as `NaN`. You can opt-in to treating all uncovered bases as having value 0.0 via `uncovered=0`.
* `fillna: float | None` -- replaces `missing` and determines how any in-bounds `NaN` values are filled after rasterization. The previous default behavior was to replace them with 0. The new default `fillna=None` is to leave them as is (which is coincidentally what pyBigWig does). Users can opt in to the old behavior with `fillna=0`, which is much more ergonomic than post-filling using numpy primitives.

Through the combination of `summary` and `uncovered`, we expose the equivalents of `mean0`, `std0`, `min0`, `max0`. As a bonus, you can insert any finite background value for uncovered bases you like.

This PR also exposes the full set of internal summary stats, including `sum`, `sum_squares`, `bases_covered` (number of bases covered by features), and `bin_covered` (fraction of the bin covered by features). The latter two are not affected by the `uncovered` argument by definition.

### Deprecation cycle

**Default behavior change**: when `fillna` is not passed, empty bins / uncovered positions are now returned as NaN instead of being filled with 0. A `DeprecationWarning` is raised on the first call without an explicit `fillna` to flag the change for existing users; pass `fillna=0` to keep prior behavior or `fillna=None` to silence the warning.

In order to emit warnings, a thin Python wrapper was introduced around the Rust `BBIReader` to detect "user did not pass `fillna`" (a distinction PyO3 can't make at the Rust boundary) and emit the `DeprecationWarning`. It can be removed in a future release when the deprecation is done.

### API changes (Python `BBIReader.values()`)

| Parameter      | Status               | Semantics                                                     | Default |
| -------------- | -------------------- | ------------------------------------------------------------- |---------|
| `uncovered`    | **new**              | Per-base value for uncovered bases used *during* the summary computation. `None` (default) excludes them (classical stats); `0.0` reproduces UCSC's `*0` stats; any finite value is a well-defined generalization. | None (exclude) |
| `fillna`       | **new**              | Post-rasterization fill for in-bounds NaN positions/bins. OOB positions are not touched by `fillna`. | None (keep NaNs) |
| `missing`      | **deprecated**       | Accepted as an alias for `fillna` with a `DeprecationWarning`.  | old default was 0.0; now opt-in with `fillna=0` |
| `oob`          | unchanged            | Fill for positions outside the chromosome's valid range.       | NaN |

## 2. UCSC parity in zoom summary interpolation (ceil + rescale normalization)

The zoom-interpolation code path now matches UCSC's `bbiSummarySlice` normalization: per-bin `bases_covered` is rounded up to the next integer, with `sum` and `sum_squares` rescaled by `ceil(bc)/bc` so that `mean = sum / bases_covered` is invariant. The `ceil` direction preserves the invariant that `bases_covered ≥ 1` whenever any data overlaps the bin (a `floor` would erase data near zoom-record boundaries). `std` is not algebraically preserved under the rescale, but matching UCSC's reported `std` requires applying the same step.

## 3. Coverage iterator

`CoverageIterator`, a 1-D sweep-line over endpoint events that converts sorted BED entries to a run-length-encoded  coverage track (matching `bedtools genomecov -bg` semantics). 

## Tests

- **31 Rust unit tests** (up from 4): `CoverageIterator` (15 cases including sort-violation, malformed-input, stress, batching), `fill_binned` (8 cases  covering every `Summary` variant plus the integer-bin-width edge case).
- **715 Python tests** (up from ~580): parity vs `pybbi` for `sum`/`mean`/ `std`/`min`/`max`/`bin_covered`/`bases_covered` across many start/end/bin combinations and both base-level and binned modes. Adds zoom-interpolation parity tests (`test_bbi_consistent_binned_zoom_bw/bb`) against`pybbi`. Plus tests for `fillna`/`uncovered` semantics, the deprecation warnings, the `arr=` parameter, invalid arguments, and fully out-of-bounds queries.
- **`pyproject.toml`**: `filterwarnings` ignores `DeprecationWarning` from `pybigtools` during the test run; deprecation tests opt back in locally via `pytest.warns(...)`.

## Migration notes for users

```python
# old
bw.values("chr1", 0, 100, bins=10, missing=-1)                # → DeprecationWarning
bw.values("chr1", 0, 100, bins=10)                            # → DeprecationWarning, NaN on empty (changed)

# new
bw.values("chr1", 0, 100, bins=10, summary="mean", uncovered=0)  # translates to "mean0"
bw.values("chr1", 0, 100, bins=10, fillna=-1)                    # = old missing=-1
bw.values("chr1", 0, 100, bins=10, fillna=None)                  # opt in, no warning
bw.values("chr1", 0, 100, bins=10, fillna=0)                     # opt back into old default
```

### CODE REVIEW

**Comment by jackh726** [comment]
I'm trying to remember - I had thought a bunch about this, and there was some reason (or some difference) in choosing `missing` versus the `fillna` parameter. I'll have to think a bit about it and try to recall. Maybe splitting into `uncovered` and `fillna` fixes this.

**Comment by nvictus** [comment]
`missing` is the name I used in `pybbi` which I now regret given the confusing semantics. Whatever we do here I may backport there.

**Comment by nvictus** [comment]
Fixed the nightly build issue @jackh726.

**Comment by nvictus** [comment]
any guidance @jackh726?

**Comment by jackh726** [comment]
Planning to take a look at this later today :) will drop by Slack and say hi!

**Comment by jackh726** [comment]
So

>bw.values("chr1", 0, 100, bins=10)                            # → DeprecationWarning, NaN on empty (changed)

I wouldn't expect a deprecation warning here: I think we should have the right defaults.

File: pybigtools/pybigtools/__init__.py [hunk 1]
```diff
@@ -1,10 +1,95 @@
...
+            if fillna is _UNSET:
+                fillna = missing
+        if fillna is _UNSET:
+            _warnings.warn(
+                "The default behavior of `values()` has changed: empty bins / "
+                "uncovered positions are now returned as NaN instead of "
+                "filled with 0. Pass `fillna=0` to keep the previous "
+                "behavior, or `fillna=None` to silence this warning. See "
+                "also `uncovered` to include uncovered bases in summary "
+                "statistic calculations.",
```
Comment (jackh726) [inline pybigtools/pybigtools/__init__.py:72]: Hmm. So, I'm assuming that the difference between `NaN` and `0` is the denominator for `mean`?

Comment (jackh726) [reply pybigtools/pybigtools/__init__.py:72]: As I said in the comment, I don't like just warning for what should be the sensible default.

If anything, we could release a patch bump that inserts a warning on *previous* versions. And then make the change in the minor bump.

File: pybigtools/tests/test_io.py [hunk 2]
```diff
@@ -88,16 +89,20 @@ def test_bigbed_write(tmpdir):
     string chrom;        "Reference sequence chromosome or scaffold"
     uint   chromStart;   "Start position in chromosome"
     uint   chromEnd;     "End position in chromosome"
+    string name;         "Name of item"
+    string name2;        "Another name"
+    float  value;        "Value"
```
Comment (jackh726) [inline pybigtools/tests/test_io.py:94]: Minor point, but this no longer a `bed3`

File: pybigtools/src/raster.rs [hunk 3]
```diff
@@ -442,6 +447,28 @@ fn fill_binned<I: Iterator<Item = Result<BBIRecord, BBIReadError>>>(
         }
     }
 
+    // Normalize bases_covered to an integer count, matching UCSC's convention.
```
Comment (jackh726) [inline pybigtools/src/raster.rs:450]: Does this need to also happen in Rust lib?

File: pybigtools/tests/test_raster.py [hunk 4]
```diff
@@ -45,213 +45,285 @@ def bb_path():
...
+    # Coverage:   1/2   1/2   3/3     2/2     1/3
+    # Sums:       1     1     6       5       3
+    # mean0:      0.5   0.5   2.0     2.5     1.0
     for path in [bw_path, bb_path]:
         x1 = pbt.open(path).values(
-            "chr1", 0, 12, bins=5, exact=True, missing=missing, summary="mean0"
-        )
-        assert np.allclose(
-            x1, [1 / 2.4, 1 / 2.4, 6 / 2.4, 5 / 2.4, 3 / 2.4], equal_nan=True
+            "chr1", 0, 12, bins=5, exact=True, uncovered=uncovered, summary="mean"
```
Comment (jackh726) [inline pybigtools/tests/test_raster.py:190]: So, no longer `mean0`, just `mean` and specify `uncovered`?

Comment (jackh726) [reply pybigtools/tests/test_raster.py:190]: Also, the diff here makes it difficult to know if the *previous* `summary=mean` matches the current or not?

Comment (jackh726) [reply pybigtools/tests/test_raster.py:190]: (and min0, etc.)

**Comment by nvictus** [comment]
> So
> 
> > bw.values("chr1", 0, 100, bins=10)                            # → DeprecationWarning, NaN on empty (changed)
> 
> I wouldn't expect a deprecation warning here: I think we should have the right defaults.

The deprecation is to signal a change in behavior to anyone depending on the old defaults so they explicit opt-in to the old behavior: `fillna=0)`. I think the new default behavior of not filling in NaNs in saner, given the semantics. 

We could change this behavior silently, assuming nobody would be affected, or be loud about it for a single release and remove the warning in the following release.
